manager/controllers/app/moduleinstance.go

```diff
@@ -75,13 +90,96 @@ func (m *ModuleManager) GetCopyDestination(item modules.DataInfo, destinationInt
 	}, nil
 }
 
-// SelectModuleInstances selects the necessary read/copy/write modules for the blueprint
+// SelectIngestModuleInstances creates the modules with their params needed to copy external data into the managed environment
+// Currently assumes that a single copy module will do the job
+func (m *ModuleManager) SelectIngestModuleInstances(item modules.DataInfo) ([]modules.ModuleInstanceSpec, error) {
+	var copySelector *modules.Selector
+	m.Log.Info("Select ingest module instances for " + item.AssetID)
+
+	// Allocate the destination storage in the relevant geo
+	var destDataStore *app.DataStore
+	var err error
+	if destDataStore, err = m.GetCopyDestination(item, item.AppInterface, item.Geo); err != nil {
+		return nil, err
+	}
+
+	// Find a copy module that is able to copy from the source to the destination
+	instances := make([]modules.ModuleInstanceSpec, 0)
+
+	// Select a module that supports COPY flow for ingest
+	actionsOnCopy := item.Actions[app.Copy]
+	if !actionsOnCopy.Allowed {
+		m.Log.Info("Ingest not allowed")
+		return instances, errors.New(actionsOnCopy.Message)
+	}
+
+	source, err := StructToInterfaceDetails(item)
+	if err != nil {
+		m.Log.Info("StructToInterfaceDetails failed for: ")
+		utils.PrintStructure(item, m.Log, "Ingest modules.DataInfo")
+
+		return instances, err
+	}
+
+	sourceDataStore := &app.DataStore{
+		Connection:         item.DataDetails.GetDataStore(),
+		CredentialLocation: utils.GetDatasetVaultPath(item.AssetID),
```

Comment:
This won't work for ingest. CredentialLocation is a part of the provided external data structure.

Reply:
You're right.  Fixed